Support expanding variables in openqa-clone-job

okurz opened the pull request:
Related progress issue: https://progress.opensuse.org/issues/182645

okurz committed:
feat: support expanding variables in openqa-clone-job

Also adds an early die on unexpanded variables in clone attempt

Related progress issue: https://progress.opensuse.org/issues/182645

asmorodskyi commented on lib/OpenQA/Script/CloneJobSUSE.pm:
```diff
@@ -28,6 +30,9 @@ sub verify_incident_repos ($url_handler, $incident_repos) {
     my @incident_urls;
     my $ua = $url_handler->{ua};
     foreach my $incident (split(/,/, $incident_repos)) {
+        die
```
I don't think it is good idea , maybe just print warning instead ? It reminds me story around https://github.com/os-autoinst/openQA/pull/6468 which showed us that it is not always good thing to decide on behalf of caller if it is make sense to continue . I still some cases where no matter that incident repos are not expanded run still make sense . For example some VRs where I simply skip use of this variables

Martchus replied:
I wouldn't mind the `die`. If we continue here we would probably just run into a hard failure later so failing fast is probably the best thing to do here.

Martchus replied:
I've just double-checked and code in `CloneJobSUSE.pm` is only executed if `--check-repos` is present.

Martchus commented on lib/OpenQA/Script/CloneJobSUSE.pm:
```diff
@@ -28,6 +30,9 @@ sub verify_incident_repos ($url_handler, $incident_repos) {
     foreach my $incident (split(/,/, $incident_repos)) {
+        die
+"URL '$incident' contains an unexpanded variable. Specify the necessary variables at the command line for expansion. See https://open.qa/docs/#_variable_expansion for details"
+          if $incident =~ /%/;
```
The `%` sign is commonly used in [URL encoding](https://en.wikipedia.org/wiki/Percent-encoding) so this check seems quite wrong.

Martchus commented on lib/OpenQA/Script/CloneJobSUSE.pm:
```diff
@@ -5,10 +5,12 @@ package OpenQA::Script::CloneJobSUSE;
 our @EXPORT = qw(detect_maintenance_update);
 
 sub collect_incident_repos ($url_handler, $settings) {
+    OpenQA::JobSettings::expand_placeholders($settings);
```
This function should probably return whether some placeholders could not be expanded. Don't do this by checking for a  `%` as mentioned in the other comment.

okurz replied:
Ok. I suggest we do
* #6662

first to ensure that the worker has no problems with the return value of expanding and then I can try to incorporate that here.

Martchus commented on lib/OpenQA/Script/CloneJobSUSE.pm:
```diff
@@ -28,6 +30,9 @@ sub verify_incident_repos ($url_handler, $incident_repos) {
     my $ua = $url_handler->{ua};
     foreach my $incident (split(/,/, $incident_repos)) {
+        die
+"URL '$incident' contains an unexpanded variable. Specify the necessary variables at the command line for expansion. See https://open.qa/docs/#_variable_expansion for details"
```
This needs a trailing `\n`, otherwise Perl prints a line number and it looks like a mistake in the program itself:
```suggestion
"URL '$incident' contains an unexpanded variable. Specify the necessary variables at the command line for expansion. See https://open.qa/docs/#_variable_expansion for details\n"
```

Martchus reviewed (changes requested):
Not repeating my comments from https://github.com/os-autoinst/openQA/pull/6654. In cases like this it would probably make sense to just open on PR.

okurz committed:
feat: improve detection of unexpanded variables in clone-job

Modify OpenQA::JobSettings::expand_placeholders to return a list of
unexpanded variables when called in list context. This allows callers to
detect missing variables without relying on fragile regex checks on the
result.

In OpenQA::Script::CloneJobSUSE, use this new capability to provide
better error messages for unexpanded variables and remove the incorrect
check for any '%' character, which was failing on percent-encoded URLs.

Related review: https://github.com/os-autoinst/openQA/pull/6655

Martchus reviewed (changes requested):
Looks generally good. I only have a few nitpicks.

okurz commented on lib/OpenQA/JobSettings.pm:
```diff
@@ -54,31 +56,39 @@ sub generate_settings ($params) {
+    }
+    catch ($e) {
+        my $err = "Error: $e";
+        return wantarray ? ($err, [sort keys %unexpanded]) : $err;
```
done

Martchus commented on lib/OpenQA/Script/CloneJobSUSE.pm:
```diff
@@ -5,10 +5,18 @@ package OpenQA::Script::CloneJobSUSE;
 our @EXPORT_OK = qw(detect_maintenance_update);
 
 sub collect_incident_repos ($url_handler, $settings) {
+    my ($error, $unexpanded) = OpenQA::JobSettings::expand_placeholders($settings);
```
Maybe a stupid question, but how will this magically fill in `%REPO_MIRROR_HOST%` (the variable mentioned in https://progress.opensuse.org/issues/182645)? I thought this variable comes from the worker config and is therefore not resolved right until the job is executed.

Martchus replied:
And for variables that can be resolved here this change is also not very useful because those variable have already been resolved at the time of scheduling (in the call you adjust in `ScheduledProducts.pm`). Or do I miss something here?

okurz replied:
Honestly I don't remember from my 2 year old changes what I fully intended but I guess you are right. openqa-clone-job running on the client side cannot expand variables that are evaluated on the worker side but we can still add support to expand *if* specified, e.g. for cloning to local workers that don't have those settings.

The previous implementation in this branch was too aggressive and would fail the clone job entirely if any variables remained unexpanded. I just pushed a change that corrects this behavior by warning instead of die.

Martchus replied:
I guess with "Specify the necessary variables at the command line for expansion." this change makes sense.
